### PR TITLE
Feat/save abtests results

### PR DESCRIPTION
**Pull request by lucas-daher**
#### What is the purpose of this pull request?
Save the results of all AB tests done.

#### What problem is this solving?
Currently the data is erased when the test is finished and we have no way of keeping track of tests performed, results etc.

#### How should this be manually tested?
Link the ab tester;
Start an ab test;
Finish the test;
Check if the data was saved.

#### Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Requires change to documentation, which has been updated accordingly.

A couple of things:
- I'm not sure if what I did was good practice (see comment in the code);
- I have no idea how to create automated tests for this. For the ab tester overall, actually. But will look into it soon.

### CODE REVIEW

File: node/clients/vbase.ts [hunk 1]
```diff
@@ -27,25 +28,43 @@ export default class VBase {
...
+    }
+    catch {
+      //does nothing, if this causes an error, probably just means resultsList.json desn't exist yet and will be created
+    }
+    finally {
+      list.push(testResults)
+      await this.save(results, testResults, ctx)
+      await this.save(list, resultsList, ctx)
+      console.log("check point111")
+    }
```
Comment (lucas-daher) [inline node/clients/vbase.ts]: Any better ideas on how to do this?

File: node/clients/vbase.ts [hunk 2]
```diff
@@ -27,25 +28,40 @@ export default class VBase {
...
+      list = JSON.parse(listFile.data.toString()) as string[]
+    }
+    catch {
+      //does nothing, if this causes an error, probably just means resultsList.json desn't exist yet and will be created
+    }
+    finally {
+      list.push(testResults)
+      await this.save(results, testResults, ctx)
+      await this.save(list, resultsList, ctx)
+    }
```
Comment (lucas-daher) [inline node/clients/vbase.ts]: Any better ideas on how to do this?

Comment (pedro823) [reply node/clients/vbase.ts]: I think the best idea is to separate the fetcher of the file in a different function, and then use something like
```js
const testResults = [...await this.fetchTestResults(ctx.account), testResultsFile]
await Promise.all([this.save(results, testResultsFile, ctx), this.save(testResults, resultsList, ctx)])
```
(Assume that you renamed `testResults` to `testResultsFile`)

That way, you:
- Have smaller, neater code
- Do both `save` calls asynchronously "at the same time"

File: node/abTest/commands/finish.ts [hunk 3]
```diff
@@ -14,9 +14,12 @@ export async function FinishAbTestForWorkspace(ctx: ColossusContext): Promise<vo
...
       await router.deleteParameters(account)
       await router.deleteWorkspaces(account)
-      let beginning = (await vbase.get(ctx.vtex)).dateOfBeginning
-      if (beginning === undefined) {
+      let beginning = ""
+      const data = await vbase.get(ctx.vtex)
+      if (!data || data.dateOfBeginning === undefined) {
         beginning = new Date().toISOString().substr(0, 16)
+      } else {
+        beginning = data.dateOfBeginning
```
Comment (tergol1) [inline node/abTest/commands/finish.ts]: **nit** WDYT about ternary operator?

```
const beginning = (!data || data.dateOfBeginning === undefined) ?
                  new Date().toISOString().substr(0, 16) :
                  data.dateOfBeginning
```

File: node/clients/vbase.ts [hunk 4]
```diff
@@ -27,25 +28,40 @@ export default class VBase {
...
       initialStageTime: initialTime,
-    } as VBaseABTestData, ctx)
+    } as VBaseABTestData, fileName, ctx)
   }
 
-  public finishABtest = async (ctx: IOContext): Promise<void> => {
+  public finishABtest = async (ctx: IOContext, results: TestResult[]): Promise<void> => {
+    
     await this.client.deleteFile(bucketName(ctx.account), fileName)
+    let testResults = 'TestResults' + results[0].ABTestBeginning + '.json'
```
[the remark this answers is not in the document]
Comment (tergol1) [reply node/clients/vbase.ts]: Good question! I think this should be checked by this point even if following the logic here this should always exists.

File: node/clients/vbase.ts [hunk 5]
```diff
@@ -27,25 +27,27 @@ export default class VBase {
...
       initialProportion: proportion,
       initialStageTime: initialTime,
-    } as VBaseABTestData, ctx)
+    } as VBaseABTestData, fileName, ctx)
   }
 
-  public finishABtest = async (ctx: IOContext): Promise<void> => {
+  public finishABtest = async (ctx: IOContext, results: TestResult[]): Promise<void> => {
     await this.client.deleteFile(bucketName(ctx.account), fileName)
+    let file = 'TestResults' + results[0].ABTestBeginning + '.json'
```
Comment (tergol1) [inline node/clients/vbase.ts]: **nit** `const file = 'TestResults' + results[0].ABTestBeginning + '.json'`

Comment (tergol1) [inline node/clients/vbase.ts]: **nit** You can make this a const in this file like `bucketName` as 
```
const testResultsFileName = (results: TestResult[]) => `TestResults${results[0].ABTestBeginning}.json`
```

File: node/clients/vbase.ts [hunk 6]
```diff
@@ -27,25 +27,27 @@ export default class VBase {
     }
   }
 
-  public save = async (data: any, ctx: IOContext) => {
+  public save = async (data: any, file: string, ctx: IOContext) => {
```
[the remark this answers is not in the document]
Comment (lucas-daher) [reply node/clients/vbase.ts]: I like the name as is for a few reasons:
- this is on a different abstraction level than the `saveFile` functions, so there should be no confusion;
- this name is consistent with the `get` function;

File: node/clients/vbase.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (lucas-daher) [reply node/clients/vbase.ts]: great point, it actually isn't.

File: node/clients/vbase.ts [hunk 7]
```diff
@@ -27,25 +28,40 @@ export default class VBase {
...
 
-  public finishABtest = async (ctx: IOContext): Promise<void> => {
+  public finishABtest = async (ctx: IOContext, results: TestResult[]): Promise<void> => {
+    
     await this.client.deleteFile(bucketName(ctx.account), fileName)
+    let testResults = 'TestResults' + results[0].ABTestBeginning + '.json'
+    let list: string[] = []
+    try {
+      const listFile = await this.client.getFile(bucketName(ctx.account), resultsList)
+      list = JSON.parse(listFile.data.toString()) as string[]
```
[the remark this answers is not in the document]
Comment (lucas-daher) [reply node/clients/vbase.ts]: TBH, I don't really know how this works, just copied a similar line from the `get` function

File: node/clients/vbase.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (tergol1) [reply node/clients/vbase.ts]: I agree with both points but I would like to think about both names also being changed. But this is just my opinion and you're free to ignore the suggestion.

**Pull request merged**